### PR TITLE
Unit tests for Microsoft.Data.SqlClient.SqlClientEventSource

### PR DESCRIPTION
**Pull request by winseros**
Hi, @DavoudEshtehari. I added the tests that check the event-counters functionality. I managed to cover most of the counters.

Also, I changed the places where the `SqlClientEventSource.Log.ActiveConnectionPoolGroupRequest(false)` and `SqlClientEventSource.Log.ActiveConnectionPoolGroupRequest(false)` calls are made, so the calls are invoked in a more "reusable" manner - closer to the place where the actual logic lives.

As there are lots of tests running in the same session and lots of the tests does affect the event-counters when run, I had to call the `DbConnectionFactory.PruneConnectionPoolGroups()` and `DbConnectionFactory.ClearAllPools()` before each test to get the underlying pooling infrastructure reset and tests reproducible. Those are internal methods, so I used reflection for that. But as an alternative, it could be done by allowing the `[Microsoft.Data.SqlClient.ManualTesting.Tests]` assembly members to access the `[Microsoft.Data.SqlClient]` internals.

I'm looking forward to your feedback and suggestions.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/Common/DbConnectionStringCommon.cs [hunk 1]
```diff
@@ -487,7 +487,7 @@ internal static ApplicationIntent ConvertToApplicationIntent(string keyword, obj
 
         internal static bool IsValidAuthenticationTypeValue(SqlAuthenticationMethod value)
         {
-            Debug.Assert(Enum.GetNames(typeof(SqlAuthenticationMethod)).Length == 6, "SqlAuthenticationMethod enum has changed, update needed");
+            Debug.Assert(Enum.GetNames(typeof(SqlAuthenticationMethod)).Length == 7, "SqlAuthenticationMethod enum has changed, update needed");
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/Common/DbConnectionStringCommon.cs]: After updating the branch with the latest source code. This change made a conflict.

Comment (winseros) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/Common/DbConnectionStringCommon.cs]: @DavoudEshtehari, that was expected. The code caused multiple tests failing so I'm not surprised someone also made the same fix as me.

I removed my changes of those lines.

**Comment by karinazhou** [comment]
Hi @winseros ,

The `[Microsoft.Data.SqlClient.ManualTesting.Tests]` assembly members are not supposed to access Microsoft.Data.SqlClient's internal methods. If you need to call those internal functions from the tests, using reflection is acceptable. 

And according to the following comment, can we use reflection to access the counters directly instead of using EventListener?
https://github.com/dotnet/SqlClient/pull/719#issuecomment-756321651

Thanks,

**Comment by winseros** [comment]
Hi, @karinazhou 

The current implementation of the `SqlClientEventSource` has a check for an active `EventListener`: 

https://github.com/DavoudEshtehari/SqlClient/blob/7ec5012ef4c36804a3250467963b2ed14c286d6e/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlClientEventSource.NetCoreApp.cs#L337-L350

I see two options here:

1. Use reflection in tests and remove the check
2. Keep using `EventListener` in tests

What's your opinion?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs [hunk 2]
```diff
@@ -689,14 +683,23 @@ public static string RetrieveValueFromConnStr(string connStr, string[] keywords)
 
         public class TraceEventListener : EventListener
         {
+            private IList<EventSource> _eventSources = new List<EventSource>();
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs]: This TraceEventListener class is also used for EventSourceTest. Is there any reason this is changed here?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs [hunk 3]
```diff
@@ -86,11 +85,6 @@ static DataTestUtility()
 
             System.Net.ServicePointManager.SecurityProtocol |= System.Net.SecurityProtocolType.Tls12;
 
-            if (TracingEnabled)
-            {
-                TraceListener = new TraceEventListener();
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs]: We need this TraceListener to capture all the traces when `TracingEnabled` flag is on in config file.

**Comment by karinazhou** [comment]
@winseros  Yes, you are right. To use reflections, this IsEnabled() condition needs to be removed. Otherwise, no points to use reflection if we need the EventListener to enable the events for counters. 
@DavoudEshtehari  Do you have any thoughts on this one?

**Comment by DavoudEshtehari** [comment]
If we remove the `IsEnabled()` condition, the counters will be calculated all the time even when we don't want to collect them. The diagnostic tools like `Event Counters` should have the lowest perf impact and this is the aim of this function and I don't recommend removing it.

**Comment by DavoudEshtehari** [comment]
After consultations with other teams, there are some perf benefits of removing the `If` statement at CPU micro-instruction level just in this case. So, go ahead and remove the `IsEnabled()` methods where counters are going to increment or decrement.

> Note: Avoid any checking before counters' manipulation, and remove the `Request()` method too.

**Comment by karinazhou** [comment]
@DavoudEshtehari  OK. I will push the changes shortly.
@winseros   Can you change the tests to use reflections instead of event listener?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs (hunk 2 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs]: The reason for the change - is the class can only enable the trace collection, but it doesn't disable it when it is no longer needed.
The result of a "non-disabled" trace collection is not visible unless there is only one trace listener all over the tests. But if at some point other listeners do appear - this one would interfere with their trace collecting.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs (hunk 3 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs]: @karinazhou, I have two points to mention here:
1. Active trace listeners influence each other. You can get tests reproducible having only a single listener active at a time. As from now, we're getting a 2nd listener, we'll need to move this one somewhere closer to the test that uses it, and explicitly handle the listener's lifetime.
2. I explored the code but did not find anything that would use this listener, so I decided it is just a dangling code. Were my findings wrong?

Comment (karinazhou) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs]: @winseros  This listener is to capture the traces for all the tests we are running. The TracingEnabled is by default turned off. This listener is for error analysis purposes. And now we are using reflection instead of a second listener for the Event Counter test so there will be no impact on different active listeners.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs (hunk 2 above)
Comment (karinazhou) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs]: Thank you for the explanation. This sounds reasonable.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs [hunk 4]
```diff
@@ -16,91 +20,250 @@ namespace Microsoft.Data.SqlClient.ManualTesting.Tests
...
-            };
+            //create a non-pooled connection
+            var stringBuilder = new SqlConnectionStringBuilder(DataTestUtility.TCPConnectionString) {Pooling = false};
 
-            using (var TraceListener = new TraceEventCounterListener())
-            {
-                OpenConnections(stringBuilder.ConnectionString);
-                stringBuilder.Pooling = false;
-                OpenConnections(stringBuilder.ConnectionString);
+            using var conn = new SqlConnection(stringBuilder.ToString());
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs]: Can we use the using block for SqlConnection here?

**Comment by karinazhou** [comment]
@winseros   I tested the unit tests locally but it failed at `EventCounter_HardConnectionsCounters_Functional` and `EventCounter_SoftConnectionsCounters_Functional` 

```
Failed Microsoft.Data.SqlClient.ManualTesting.Tests.EventCounterTest.EventCounter_HardConnectionsCounters_Functional [7 ms]
  Error Message:
   Assert.Equal() Failure
Expected: 0
Actual:   1
  Stack Trace:
     at Microsoft.Data.SqlClient.ManualTesting.Tests.EventCounterTest.EventCounter_HardConnectionsCounters_Functional() in \EventCounterTest\src\Microsoft.Data.SqlClient\tests\ManualTests\TracingTests\EventCounterTest.cs:line 33

  Failed Microsoft.Data.SqlClient.ManualTesting.Tests.EventCounterTest.EventCounter_SoftConnectionsCounters_Functional [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: 0
Actual:   1
  Stack Trace:
     at Microsoft.Data.SqlClient.ManualTesting.Tests.EventCounterTest.EventCounter_SoftConnectionsCounters_Functional() in \EventCounterTest\src\Microsoft.Data.SqlClient\tests\ManualTests\TracingTests\EventCounterTest.cs:line 58
```
---------------------------------------------------------------------------------------------------------------------

It looks like a single run of either `EventCounter_HardConnectionsCounters_Functional` or `EventCounter_SoftConnectionsCounters_Functional` passing as expected 🤔

**Comment by winseros** [comment]
@karinazhou I still work on that. After we removed the `IsEnabled()` check, other tests started affecting counters. The `EventCounterTest` tries to do a cleanup before a run, but the cleanup won't help if any test left a dangling open `SqlConnection`.

Currently, I'm looking for tests that do leave open `SqlConnection` and deal with them - mostly wrap them into `using`.

**Comment by karinazhou** [comment]
@winseros  It sounds good. Please let me know if you need any help on this. I can look into this too.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs (hunk 3 above)
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs]: According to the recent changes and using reflection, this change is not directly related to the event counter, and it'd better be wrapped up in a separate PR for further discussions.
Feel free to create a new PR  and move it there with its related explanations, please.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs [hunk 5]
```diff
@@ -16,91 +15,247 @@ namespace Microsoft.Data.SqlClient.ManualTesting.Tests
...
+        {
+            //create a non-pooled connection
+            var stringBuilder = new SqlConnectionStringBuilder(DataTestUtility.TCPConnectionString) {Pooling = false};
+
+            using var conn = new SqlConnection(stringBuilder.ToString());
+
+            //initially we have no open physical connections
+            Assert.Equal(0, SqlClientEventSourceProps.ActiveHardConnections);
+            Assert.Equal(0, SqlClientEventSourceProps.NonPooledConnections);
+
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs:36]: Can you also add the check for 
`_hardConnectsCounter - _hardDisconnectsCounter == _activeHardConnectionsCounter` after we open and close non-pooled connection ?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs [hunk 6]
```diff
@@ -16,91 +15,247 @@ namespace Microsoft.Data.SqlClient.ManualTesting.Tests
...
+                Assert.Equal(0, SqlClientEventSourceProps.ActiveHardConnections);
+                Assert.Equal(0, SqlClientEventSourceProps.ActiveSoftConnections);
+                Assert.Equal(0, SqlClientEventSourceProps.PooledConnections);
+                Assert.Equal(0, SqlClientEventSourceProps.NonPooledConnections);
+                Assert.Equal(0, SqlClientEventSourceProps.ActiveConnectionPools);
+                Assert.Equal(0, SqlClientEventSourceProps.ActiveConnections);
+                Assert.Equal(0, SqlClientEventSourceProps.FreeConnections);
+
+                conn.Open();
+
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs:78]: Similar to the test case above this one. Can you also add the check for 
`_hardConnectsCounter - _hardDisconnectsCounter == _activeHardConnectionsCounter`
and 
`_softConnectsCounter - _softDisconnectsCounter == _activeSoftConnectionsCounter`
after we open and close and reopen connection?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionPoolTest/ConnectionPoolTest.cs [hunk 7]
```diff
@@ -178,20 +178,23 @@ private static void ReclaimEmancipatedOnOpenTest(string connectionString)
             string newConnectionString = (new SqlConnectionStringBuilder(connectionString) { MaxPoolSize = 1 }).ConnectionString;
             SqlConnection.ClearAllPools();
 
-            InternalConnectionWrapper internalConnection = CreateEmancipatedConnection(newConnectionString);
+            using var outerConnection = new SqlConnection(newConnectionString);
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionPoolTest/ConnectionPoolTest.cs]: Please revert the changes in this `ReclaimEmancipatedOnOpenTest` test because it is designed to trigger the reclaim behavior without explicitly close the connection on purpose. Otherwise, when the new connection is created, the internalConnection won't be reclaimed.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionPoolTest/ConnectionPoolTest.cs [hunk 8]
```diff
@@ -280,12 +283,5 @@ private static InternalConnectionWrapper ReplacementConnectionUsesSemaphoreTask(
 
             return internalConnection;
         }
-
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionPoolTest/ConnectionPoolTest.cs]: Please bring this back which is needed for the `ReclaimEmancipatedOnOpenTest`.

**Comment by karinazhou** [comment]
Hi @winseros ,  

Here are somethings needed to be updated before we can give this PR a final review:

1. The `using` statements you added are not supported in the language version we use. This creates a lot of building errors. Please change it to the old `using` block style.

2. The last commit doesn't seem to solve the dangling connection from other tests. Since we are going to keep some of the SqlConnections without explicitly calling Close() or Dispose() methods such as `ReclaimEmancipatedOnOpenTest`, it cannot guarantee the exact number of some of the counters even we call SqlConnection.ClearAllPools() and PruneConnectionPoolGroups(). 

One workaround is to compare the counters before and after SqlConnection.Open() or SqlConnection.Close(). The expected behavior will be to increase / decrease the counter by 1. This can eliminate the impact from other tests running before `EventCounterTest`.

3. Please revert the changes in `ReclaimEmancipatedOnOpenTest` according to my comments. 

4. Please add the following code back in src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs

```
if (TracingEnabled)	
{	
                TraceListener = new TraceEventListener();
}
```

5. Please revert the TraceEventListener change in this PR which is not relevant to EventCounterTest. You can create a separate PR in SqlClient/master branch to include this change.

Thanks,

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionPoolTest/ConnectionPoolTest.cs (hunk 7 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionPoolTest/ConnectionPoolTest.cs]: Done.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs (hunk 6 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs:78]: Done

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs (hunk 5 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs:36]: Done

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionPoolTest/ConnectionPoolTest.cs (hunk 8 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionPoolTest/ConnectionPoolTest.cs]: Done

**Comment by winseros** [comment]
Hi, @karinazhou. 

#### The current status:

1. Done. I reverted the `using`  statements to the classic form. I hope I did not miss anyone.
2. I'll share my thoughts on this one below.
3. Done. Reverted the `ReclaimEmancipatedOnOpenTest`
4. Done. Reverted the `DataTestUtility`
5. Done. Reverted the `TraceListener`

#### Regarding the dangling connections. 

When I started to work on tests, I tried to remember event counters' state before a test and check the changes after. The approach was working until I began to check the connection-pool related stuff. The connection pool counters appeared to be too unpredictable and depend on the pool state at the moment of the test. So the only way I see to get the `EventCountersTest` reproducible - is to have the dangling connections killed by one means or another.

So far, the `EventCounterTest` works fine (at least on my env) with all the tests except two. Those two are `ConnectionPoolTest` (its `ReclaimEmancipatedOnOpenTest` method) and `ConnectionResiliencySPIDTest`.

I think at this moment I'll ask for @DavoudEshtehari `s help. I'm not sure how the internals should work, but I feel those two tests strike event counters out of balance.

1. The `ReclaimEmancipatedOnOpenTest` makes a previous `SqlConnection` be collected by the GC and a new one to take its resources. But when the first `SqlConnection` gets GC'ed, the counters don't decrement, and when a new one opens, counters do increment, what looks superfluous.
2. The `ConnectionResiliencySPIDTest` triggers the `DbConnectionPool.ReplaceConnection()`. I'm not sure I can provide the precise info right now, but it seems that counters are out of balance after this call.

@DavoudEshtehari could you please have a look and share your thoughts on those two?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/DateTimeVariantTest.cs [hunk 9]
```diff
@@ -345,7 +347,7 @@ private static void TestSqlDataReaderParameterToTVP_Variant(object paramValue, s
                     xsql(conn, string.Format("create type dbo.{0} as table (f1 sql_variant)", tvpTypeName));
 
                     // Send TVP using SqlDataReader.
-                    SqlConnection connInput = new SqlConnection(s_connStr);
+                    using SqlConnection connInput = new SqlConnection(s_connStr);
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/DateTimeVariantTest.cs]: missing using block

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/DateTimeVariantTest.cs [hunk 10]
```diff
@@ -727,7 +729,7 @@ private static void SqlBulkCopySqlDataReader_Type(object paramValue, string expe
                     }
                     xsql(conn, string.Format("insert into {0}(f1) values(CAST('{1}' AS {2}));", bulkCopySrcTableName, value, expectedBaseTypeName));
 
-                    SqlConnection connInput = new SqlConnection(s_connStr);
+                    using SqlConnection connInput = new SqlConnection(s_connStr);
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/DateTimeVariantTest.cs]: missing using block

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs (hunk 4 above)
Comment (karinazhou) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs]: missing using block

**Comment by DavoudEshtehari** [comment]
That's right @winseros, we don't apply the GC effect on the connection pool because that would eliminate the benefits of having a pooler by causing another round trip to the server.
The `_reclaimedConnectionsCounter` correctly shows the number of reclaimed connections in total. We reclaim any released objects that were not closed before creating any new objects. Some of the event counters may show an inaccurate number for an unknown duration before reclaiming those objects, and we shouldn't increment a counter if we didn't decrease it earlier.

Generally, to have more precise tests and avoid conflicts with other running tests, the following solutions would be useful:
- Enforce the tests to run sequentially instead of parallel by xUnit.
- Isolate the running tests by using one or more separate app domains.

#### Related links:
- [Remove connections](https://docs.microsoft.com/en-us/sql/connect/ado-net/sql-server-connection-pooling?view=sql-server-ver15#remove-connections)
- [AppDomain](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain?view=net-5.0)

__PS__: Default `xUnit`'s behaviour has set to run sequentially per assembly by a single thread.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/DateTimeVariantTest.cs (hunk 9 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/DateTimeVariantTest.cs]: Done

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/DateTimeVariantTest.cs (hunk 10 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ParameterTest/DateTimeVariantTest.cs]: Done

**Comment by winseros** [comment]
@DavoudEshtehari I'm not sure `AppDomain` will work, assuming all the stuff runs on `.Net Core`. As far as I know, `AppDomains` are out of support there.

I found only two options that really work:
1. Move the `EventCounterTest` to a dedicated project. This way `xUnit` runs it isolated from the other tests.
2.  Make the `EventCounterTest` run before the others. It will require a custom [xUnit TestCaseOrderer](https://github.com/xunit/xunit/blob/6bc9543856f17cd69ec96e63307529edec2c60cd/src/xunit.v3.core/TestCaseOrdererAttribute.cs) added to the test project, at the assembly level. (Here is the [default orderer](https://github.com/xunit/xunit/blob/3aeb38c4eeb0cc8053119fdd638da943afa0d5a9/src/xunit.v3.core/Sdk/v3/Utility/DefaultTestCaseOrderer.cs), by the way). Apparently that will change the current run order of all the tests, and I can't say right now what and how much it will affect.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs (hunk 4 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs]: Done

**Comment by DavoudEshtehari** [comment]
@winseros Could you explain more about the approach of studying the variation of counters values' problem by adding some samples you encountered with them to more examine before choosing other options?

**Comment by winseros** [comment]
Hi, @DavoudEshtehari, sorry for answering with such a delay.

I hope I understood the right way what you mean and giving a meaningful reply. 

Here is what I'm doing to reproduce the issue.

1. Ensure the `EventCounterTest` works alone. Select it in the TestExplorer and run. Here is the result - the test passes:
   image
2. Now run the test together with the `ConnectionPoolTest`. The `ConnectionPoolTest` runs first due to the `xUnit` default ordering strategy and the `EventCounterTests` fails after.
   image
   Putting breakpoints at the failing assertions will give a picture of what exactly went wrong. and putting breakpoints inside the `SqlClientEventSource` which way did it happen.
3. Repeat the experiment, now with the `ConnectionResiliencySPIDTest`, and see the similar results. 
   image
   Now breakpoints inside the `DbConnectionPool.ReplaceConnection()` would be useful to understand what happens.

Please let me know if I can provide anything else that would help.

**Comment by karinazhou** [comment]
@winseros   Thank you for showing the issue in detail. I think what @DavoudEshtehari says is about why this method doesn't work in your case:

>When I started to work on tests, I tried to remember event counters' state before a test and check the changes after. The approach was working until I began to check the connection-pool related stuff. The connection pool counters appeared to be too unpredictable and depend on the pool state at the moment of the test. 

Since you have cleared the connection pools before running these three event counter test cases and the xunit runs the tests sequentially based on our setting, we are expecting the before-after check to reflect the changes correctly. Any reason we can't do it in this way?

**Comment by DavoudEshtehari** [comment]
Hi @winseros, thank you for the answer.
I made a quick check to see the effect of checking the differentiation of values in `EventCounter_SoftConnectionsCounters_Functional`.
You can find its result here:

image

1. Store the counter values before opening a connection:
image

2. Study the expected values according to the stored values in the first step:
- After `open` a connection:
image
- After `close` the connection:
image
- After `open` the second connection:
image

**Comment by winseros** [comment]
@DavoudEshtehari I rewrote the tests as you pointed, and they started working! Thanks a lot!

**Review by karinazhou — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs [hunk 11]
```diff
@@ -2,11 +2,10 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System.Collections.Generic;
-using System.Diagnostics.Tracing;
-using System.Linq;
-using System.Threading;
-using System.Threading.Tasks;
+using System;
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs]: Unnecessary `using`.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs [hunk 12]
```diff
@@ -16,91 +15,295 @@ namespace Microsoft.Data.SqlClient.ManualTesting.Tests
...
+            var stringBuilder = new SqlConnectionStringBuilder(DataTestUtility.TCPConnectionString) {Pooling = false};
+
+            var ahc = SqlClientEventSourceProps.ActiveHardConnections;
+            var npc = SqlClientEventSourceProps.NonPooledConnections;
+
+            using (var conn = new SqlConnection(stringBuilder.ToString()))
+            {
+                //initially we have no open physical connections
+                Assert.Equal(ahc, SqlClientEventSourceProps.ActiveHardConnections);
+                Assert.Equal(npc, SqlClientEventSourceProps.NonPooledConnections);
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs]: These assessments are not necessary. You can skip the beginning assessments for other tests too.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs [hunk 13]
```diff
@@ -16,91 +15,295 @@ namespace Microsoft.Data.SqlClient.ManualTesting.Tests
...
+                Assert.Equal(SqlClientEventSourceProps.ActiveHardConnections,
+                    SqlClientEventSourceProps.HardConnects - SqlClientEventSourceProps.HardDisconnects);
+                Assert.Equal(SqlClientEventSourceProps.ActiveSoftConnections,
+                    SqlClientEventSourceProps.SoftConnects - SqlClientEventSourceProps.SoftDisconnects);
+
+                conn.Close();
+
+                //when the connection gets closed, the real physical connection gets returned to the pool
+                Assert.Equal(ahc + 1, SqlClientEventSourceProps.ActiveHardConnections);
+                Assert.Equal(asc + 0, SqlClientEventSourceProps.ActiveSoftConnections);
```
Comment (karinazhou) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs]:
```suggestion
                Assert.Equal(asc, SqlClientEventSourceProps.ActiveSoftConnections);
```

Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs]: Done

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs (hunk 12 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs]: Done

File: src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs (hunk 11 above)
Comment (winseros) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/TracingTests/EventCounterTest.cs]: Done

**Comment by DavoudEshtehari** [comment]
Thank you @winseros, for the contribution.
